src/main/java/org/opensearch/knn/index/codec/Function5Arity.java

```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec;
+
+@FunctionalInterface
+public interface Function5Arity<S, T, U, V, X, R> {
```

Comment:
This is going to solve the immediate problem where we need five params and return VectorFormat Class . These are two draw back on this interface.

1. If tomorrow I need sixth Param , I need to create separate interface with six param .
2. On every Field Vector format class , we are going to write like the way it is getting written now 

```
(maxConnm, beamWidth) -> new Lucene99HnswVectorsFormat(maxConnm, beamWidth),
            (maxConnm, beamWidth, confidenceInterval, bits, compress) -> new Lucene99HnswScalarQuantizedVectorsFormat(
                maxConnm,
                beamWidth,
                1,
                bits,
                compress,
                confidenceInterval,
                null
            `)`
```

We can make this extensible like this 

Create a class like this 

```
public class KnnVectorsFormatParams {
    private final int maxConnections;
    private final int beamWidth;
    private final float confidenceInterval;
    private final int bits;
    private final boolean compressFlag;

    public KnnVectorsFormatParams(int maxConnections, int beamWidth, float confidenceInterval, int bits, boolean compressFlag) {
        this.maxConnections = maxConnections;
        this.beamWidth = beamWidth;
        this.confidenceInterval = confidenceInterval;
        this.bits = bits;
        this.compressFlag = compressFlag;
    }

    public int getMaxConnections() {
        return maxConnections;
    }

    public int getBeamWidth() {
        return beamWidth;
    }

    public float getConfidenceInterval() {
        return confidenceInterval;
    }

    public int getBits() {
        return bits;
    }

    public boolean isCompressFlag() {
        return compressFlag;
    }
}

```

Modify Validate Method 

```
private final Function<KnnVectorsFormatParams, KnnVectorsFormat> quantizedVectorsFormatSupplier;

private KnnVectorsFormat validateAndApplyQuantizedVectorsFormatForLuceneEngine(
        final Map<String, Object> params,
        final String field,
        final int maxConnections,
        final int beamWidth
    ) {

    if (params.get(METHOD_ENCODER_PARAMETER) == null) {
        return null;
    }

    if (!(params.get(METHOD_ENCODER_PARAMETER) instanceof MethodComponentContext)) {
        return null;
    }

    MethodComponentContext encoderMethodComponentContext = (MethodComponentContext) params.get(METHOD_ENCODER_PARAMETER);
    if (!ENCODER_SQ.equals(encoderMethodComponentContext.getName())) {
        return null;
    }

    Map<String, Object> sqEncoderParams = encoderMethodComponentContext.getParameters();
    Float confidenceInterval = getConfidenceInterval(sqEncoderParams);
    int bits = getBits(sqEncoderParams);
    boolean compressFlag = getCompressFlag(sqEncoderParams);

    log.debug(
        "Initialize KNN vector format for field [{}] with params [{}] = \"{}\", [{}] = \"{}\", [{}] = \"{}\", [{}] = \"{}\", [{}] = \"{}\"",
        field, MAX_CONNECTIONS, maxConnections, BEAM_WIDTH, beamWidth, LUCENE_SQ_CONFIDENCE_INTERVAL, confidenceInterval, LUCENE_SQ_BITS, bits, LUCENE_SQ_COMPRESS, compressFlag
    );

    KnnVectorsFormatParams formatParams = new KnnVectorsFormatParams(maxConnections, beamWidth, confidenceInterval, bits, compressFlag);
    return quantizedVectorsFormatSupplier.apply(formatParams);
}








```


and Now 

This is simplified like this 

```
public class KNN990PerFieldKnnVectorsFormat extends BasePerFieldKnnVectorsFormat {
    public KNN990PerFieldKnnVectorsFormat(final Optional<MapperService> mapperService) {
        super(
            mapperService,
            Lucene99HnswVectorsFormat.DEFAULT_MAX_CONN,
            Lucene99HnswVectorsFormat.DEFAULT_BEAM_WIDTH,
            () -> new Lucene99HnswVectorsFormat(),
            params -> new Lucene99HnswVectorsFormat(
                params.getMaxConnections(),
                params.getBeamWidth()
            ),
            params -> new Lucene99HnswScalarQuantizedVectorsFormat(
                params.getMaxConnections(),
                params.getBeamWidth(),
                1,
                params.getBits(),
                params.isCompressFlag(),
                params.getConfidenceInterval(),
                null
            )
        );
    }
```

With this , 
1. we won't end up creating multiple function inteface , we can just modify param class to include new param.
2. It's cleaner and more readble way.

Reply:
The idea of having a params class is really good. One thing I would like to add here is we should have a base class of parms of HNSW and then for different format we should have child classes.

I would also like to see how we can abstract the validation logic too in those classes to ensure maximum code reusability.

Reply:
Thanks for calling it out. Based on the suggestions provided, I created a base class for Params and a separate child class for quantizedformat params. Please take a look

Reply:
since you have separated the params out why do we still need this interface?

Reply:
not required. I removed it now